sematic/resolvers/log_streamer.py

```diff
@@ -201,39 +300,56 @@ def ingested_logs(
         # true when the problem is something with the "normal" logging mechanisms, like
         # a failure to upload the logs to remote. Having *some* way to see what the code
         # was doing before it died will be essential.
-        _tail_log_file(file_path, max_tail_bytes)
+        _tail_log_file(file_path)
+        if read_file_descriptor is not None:
+            os.close(read_file_descriptor)
+        if write_file_descriptor is not None:
+            os.close(write_file_descriptor)
 
 
-def _tail_log_file(file_path, max_tail_bytes, print_func=None):
-    """Print the last lines of the log file.
+def _send_signal_or_kill(pid: int, signal_num: int, timeout_seconds: int):
+    """Send the signal to the given pid. If not exited by timeout, send SIGKILL
 
-    The code will quickly traverse to the correct file location rather than reading
-    through the whole log.
+    Parameters
+    ----------
+    pid:
+        The pid of the process to kill
+    signal_num:
+        The initial signal to send. Will be sent repeatedly until the process
+        terminates or the timeout occurs
+    timeout_seconds:
+        The maximum time to wait before sending a SIGKILL
     """
+    try:
+        started = time.time()
+        while time.time() - started < timeout_seconds:
+            os.kill(pid, signal_num)
+            wait_result = os.waitpid(pid, os.WNOHANG)
+            if wait_result is None:
+                return
+            time.sleep(0.1)
+        os.kill(pid, signal.SIGKILL)
+        os.waitpid(pid, 0)
+    except ProcessLookupError:
+        return  # process already gone
```

Comment:
If there is an error in this child log streamer, then the logs will be missing, with no indication in the worker process regarding what has happened. How about reporting this case using a custom exit code for the worker?

Reply:
Or at least a log message.

Reply:
My understanding is that this logged message will not be collected due to the log collection process being dead. So we need another solution for the "who watches the watchers" problem. That's why I suggested a custom exit code.

Reply:
Added a check on the exit code of the streamer, and raise a `RuntimeError` if it is non-zero